lib/AST/ASTMangler.cpp

```diff
@@ -2506,17 +2527,47 @@ void ASTMangler::appendProtocolName(const ProtocolDecl *protocol,
     appendDeclName(protocol);
 }
 
-const clang::NamedDecl *ASTMangler::getClangDeclForMangling(const ValueDecl *vd) {
-  auto namedDecl =  dyn_cast_or_null<clang::NamedDecl>(vd->getClangDecl());
+const ASTMangler::ClangDeclForManglingContainer
+ASTMangler::getClangDeclForMangling(const ValueDecl *vd) {
+  auto namedDecl = dyn_cast_or_null<clang::NamedDecl>(vd->getClangDecl());
   if (!namedDecl)
     return nullptr;
-  
+
   // Use an anonymous enum's enclosing typedef for the mangled name, if
   // present. This matches C++'s rules for linkage names of tag declarations.
   if (namedDecl->getDeclName().isEmpty())
-    if (auto *tagDecl = dyn_cast<clang::TagDecl>(namedDecl))
-      if (auto *typedefDecl = tagDecl->getTypedefNameForAnonDecl())
+    if (auto *tagDecl = dyn_cast<clang::TagDecl>(namedDecl)) {
+      if (auto enumDecl = dyn_cast<clang::EnumDecl>(tagDecl)) {
+        // CF_OPTION derived types are different in Objective-C vs Objective-C++
+        // due different definitions in the SDK based on if __cplusplus is
+        // defined. The linkage name of these CF_OPTION derived types must be
+        // consistent regardless of if cxx interop is turned on or not,
+        // otherwise modules compiled with cxx interop cannot interoperate with
+        // those compiled without it.
+        if (auto typedefType = dyn_cast<clang::TypedefType>(
+                enumDecl->getIntegerType().getTypePtr());
+            typedefType && vd->getASTContext().LangOpts.EnableCXXInterop) {
+          for (auto *attr : typedefType->getDecl()
+                                ->specific_attrs<clang::AvailabilityAttr>()) {
+            if (attr->getPlatform()->getName() == "swift") {
```

Comment:
This  set of nested if's looks kinda similar to the previous. Could they be merged into a common helper? If you dont think it makes things neater then dont worry about it.